### ISSUE TITLE
Some observations while using fast_dp for image classification

### ISSUE THREAD

**Issue by prahaladhchandrahasan** [awslabs/fast-differential-privacy#46]
I tried out the fast_dp library on a different dataset and was truly impressed with the almost 50% reduction in training time and the use of significantly lesser GPU memory for private optimization. But I noticed the following  behavior while trying out fast_dp:

### Privacy Accountant Behavior

- I set a **target privacy budget** of epsilon = 4.0 for the entire training process.
- However, the implemented RDP accountant ([Privacy Engine Code](https://github.com/awslabs/fast-differential-privacy/blob/main/fastDP/privacy_engine.py#L401)) returned a final epsilon value of **2.2**.
- This discrepancy requires further clarification.
- I then set the epsilon value to 100.0 and the the final epsilon returned by the accountant is **49.26**.
- We can observe that the final epsilon returned by the implemented RDP privacy accountant is half the target epsilon set by the user.
- For epsilon=100 I still got the same utility as the one with epsilon = 4.0 which is counterintiutive.

### Gradient Clipping Effect

- Normally, the model accuracy should vary with changes to the **maximum L2 norm** for gradient clipping.
- Despite adjusting the max L2 norm, there was no noticeable effect on model accuracy.

### Noise Multiplier Calculation

- Instead of pre-calculating the noise multiplier, I directly provided the **target epsilon** as **4.0** and **max_grad norm** as **40** compared to **4** to the `PrivacyEngine` class.
- The final epsilon still remained **2.2** with the same utility, which shows that there is no effect of adjusting the **max_grad norm** parameter.

I have documented all my observations in this ([Repo](https://github.com/prahaladhchandrahasan/FAST_DP)) and would like to know the reason behind my observations.

**Comment by woodyx218** [comment]
After checking your [notebook](https://github.com/prahaladhchandrahasan/FAST_DP/blob/main/Fast_DP/PathMNIST_AWS_Fast_DP.ipynb), I found the issues is not related to DP. The issue is your dataloader: you are using a logical/physical batch size 512 in "train_dl =DataLoader(train_ds,512,shuffle = True,num_workers = 4, pin_memory = True)" but you mistakenly set physical batch size to 128. This leads to a correct noise multiplier but a wrong number of steps (it shows 176 steps per epoch but could be 4X larger if your physical batch size is actually 128). That's why the total epsilon is smaller than pre-defined. Same issue applies to epsilon=100 notebook, so maybe avoid hard-coding the variables like batch size.

On a side note, we recommend to use the privacy engine to compute the noise multiplier, not to pre-calculate it. I will modify the code a bit in the next few days to better display the privacy budget during training.

**Comment by woodyx218** [comment]
As for the gradient clipping, there is no issue with our implementation because the default clipping_fn in privacy_engine.py is "automatic". See Line 200 where we commented "# max_grad_norm does not matter in automatic clipping". This is a new technique from [Automatic Clipping: Differentially Private Deep Learning Made Easier and Stronger](https://papers.nips.cc/paper_files/paper/2023/file/8249b30d877c91611fd8c7aa6ac2b5fe-Paper-Conference.pdf) that normalizes instead of clips the per-sample gradient, so we need not to tune the clipping threshold as a hyperparameter. If you set clipping_fn to "Abadi", it is expected that max L2 norm will be effective.

**Comment by prahaladhchandrahasan** [comment]
Thank you for the swift response 
I had followed the following tutorial from https://github.com/awslabs/fast-differential-privacy/blob/main/examples/image_classification/CIFAR_TIMM.py where the noise multiplier was calculated  from the Opacus function. 

In the tutorial I see a variable` n_acc_step`s which was calculated as `args.bs // args.mini_bs` . Are you implying that this should this ratio always be set to 1 so that gradients are updated after every batch? If yes then why do we need two variables `bs` and `mini_bs`?

Side note : I tried using the library today and found the following error during import. 

<img width="1181" alt="Screenshot 2025-01-08 at 12 42 32 PM" src="https://github.com/user-attachments/assets/98dee0ea-462e-4122-a814-aec301366502" />

**Comment by woodyx218** [comment]
n_acc_steps has nothing to do with privacy and can be any number. We need mini_bs only if bs is too large to fit in memory. Note that n_acc_steps is the frequncy that you actually updates the step counts in the privacy engine. Say mini_bs=1 and bs=5, then if you train for 100 mini-batches, you would only call optimizer.step() 20 times and only 20 (not 100) is relevant to privacy accounting.

We are going through some major issues so please try an earlier version of this repo.

**Comment by prahaladhchandrahasan** [comment]
Thanks for the clarification. Then I guess there is an incorrect  implementation in the [examples/image_classification/CIFAR_TIMM.py](https://github.com/awslabs/fast-differential-privacy/blob/main/examples/image_classification/CIFAR_TIMM.py) as in this tutorial in [line 35](https://github.com/awslabs/fast-differential-privacy/blob/bf8cd1b8f19a0ba85123f293a26119391a0beda9/examples/image_classification/CIFAR_TIMM.py#L35C5-L35C72) you have defined `n_acc_steps` as `args.bs // args.mini_bs` and in the training loop in [line 97](https://github.com/awslabs/fast-differential-privacy/blob/bf8cd1b8f19a0ba85123f293a26119391a0beda9/examples/image_classification/CIFAR_TIMM.py#L97) where we have condition in which the `optimizer.step()` or backpropagation takes place when `(batch_idx + 1) % n_acc_steps == 0)` . This controls the the number of batches after which gradient updates takes place I observe empirically that if this ratio is not `1` then the entire training budget is not being utilized. 

I have also tested this out independently  where I have defined `n_acc_steps = BATCH_SIZE // MAX_PHYSICAL_BATCH_SIZE` where I have tested with `n_acc_steps` with values 1 and 4 for one epoch and noted that for  `n_acc_steps = 1` the final epsilon was `3.9979488851682903` but when `n_acc_steps = 4` the final epsilon was `3.3960610351956104`.  The notebooks for the same are[ here ](https://github.com/prahaladhchandrahasan/FAST_DP/tree/main/n_acc_steps)

**Comment by woodyx218** [comment]
I think the code is correct. The confusion may rise because it is wrong "backpropagation takes place when (batch_idx + 1) % n_acc_steps == 0)". In this example, backpropagation happens every iteration and generates the grad, but this grad is not used until n_acc_steps. 

I see in [your notebook](https://github.com/prahaladhchandrahasan/FAST_DP/blob/main/n_acc_steps/PathMNIST_AWS_Fast_DP_n_acc_steps__4.ipynb) the physical batch size is still not consistent to train_dl physical batch size. Please check how gradient accumulation works.

**Comment by prahaladhchandrahasan** [comment]
I still am a bit confused. In my notebook, I have used the ` MAX_PHYSICAL_BATCH_SIZE` only for calculating the `n_acc_steps` and have not used it in the privacyengine or at any other place. Could you point out where exactly I am going wrong?

**Comment by woodyx218** [comment]
As I said, the error is in train_dl, which takes a physical bs 512 but you also set physical bs 128.

**Comment by prahaladhchandrahasan** [comment]
Thank you for your patient reply. I guess there is some miscommunication from my side and as a final step:
I have commented out the `if ((batch_idx + 1) % n_acc_steps == 0) or ((batch_idx + 1) == len(train_dl)):` while keeping the physical bs to 128. In this case I am getting an epsilon of `3.9979488851682903` .

I have implemented  notebook here : https://colab.research.google.com/drive/159Eiq6fFPtnzqw4pV-UBuZZ1d6pgE95i?usp=sharing

**Comment by woodyx218** [comment]
No problem. To clarify, train_dl only takes physical bs as input and privacy engine only takes logical bs as input. Our code does support grad accumulation and the issue is not DP-related.
